rafay/resource_eks_pod_identity.go

```diff
@@ -0,0 +1,437 @@
+	return &schema.Resource{
+		CreateContext: resourceEksPodIdentityCreate,
+		ReadContext:   resourceEksPodIdentityRead,
+		UpdateContext: resourceEksPodIdentityUpdate,
+		DeleteContext: resourceEksPodIdentityDelete,
+		// Importer: &schema.ResourceImporter{
+		// 	State: resourceEksPodIdentityImport,
+		// },
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"spec": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: podIdentityAssociationsFields(),
+				},
+				MinItems: 1,
+				MaxItems: 1,
+			},
+			"metadata": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: metadataFields(),
+				},
+				MinItems: 1,
+				MaxItems: 1,
+			},
+		},
+		//		Description:   resourceEksPodIdentityDescription,
+	}
+}
+
+func metadataFields() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"cluster_name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"project_name": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+	}
+
+	return s
+}
+
+func resourceEksPodIdentityCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	return resourceEksPodIdentityUpsert(ctx, d, m)
+}
+
+func resourceEksPodIdentityUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	log.Printf("pod identity update starts")
+	tflog := os.Getenv("TF_LOG")
+	if tflog == "TRACE" || tflog == "DEBUG" {
+		ctx = context.WithValue(ctx, "debug", "true")
+	}
+
+	var podIdentity []*IAMPodIdentityAssociation
+	metadata := &Metadata{}
+	//rawConfig := d.GetRawConfig()
+
+	if v, ok := d.Get("spec").([]interface{}); ok && len(v) > 0 {
+		podIdentity = expandIAMPodIdentityAssociationsConfig(v)
+	} else {
+		return diag.FromErr(fmt.Errorf("spec not specified"))
+	}
+
+	if v, ok := d.Get("metadata").([]interface{}); ok && len(v) > 0 {
+		in := v[0].(map[string]interface{})
+		metadata.clusterName = in["cluster_name"].(string)
+		metadata.projectName = in["project_name"].(string)
+	}
+
+	resp, err := project.GetProjectByName(metadata.projectName)
+	if err != nil {
+		log.Printf("project does not exist, error %s", err.Error())
+		return diag.FromErr(err)
+	}
+	p, err := project.NewProjectFromResponse([]byte(resp))
+	if err != nil {
+		return diag.FromErr(err)
+	} else if p == nil {
+		d.SetId("")
+		return diags
+	}
+	project_id := p.ID
+
+	cluster_resp, err := cluster.GetCluster(metadata.clusterName, project_id, "terraform")
+	if err != nil {
+		log.Printf("imported cluster was not created, error %s", err.Error())
+		return diag.FromErr(err)
+	}
+
+	auth := config.GetConfig().GetAppAuthProfile()
+	uri := fmt.Sprintf("/edge/v1/projects/%s/edges/%s/podidentity", project_id, cluster_resp.ID)
+
+	// payload, err := json.Marshal(podIdentity)
+	// if err != nil {
+	// 	return diag.FromErr(err)
+	// }
+
+	log.Printf("payload response : %s", podIdentity[0].Namespace)
```

Comment:
@gopikrishna-rafay Can you add array length check?

Reply:
Done